### PR TITLE
os: always use pcLinkToDir as kind for links to directories

### PR DESCRIPTION
**Pull request by rudis**
Only the OS-specific case in walkDir() handled that correctly. Adapt the
general case and handle it in getFileInfo().

I tried to pass nil instead of "" but then nim complained:

```
Error: ambiguous call; both system.==(x: pointer, y: pointer) and system.==(x: seq[T], y: seq[T]) match for: (nil, nil)
```

I don't know enough nim to fix this so I passed "" instead.

### CODE REVIEW

File: lib/pure/os.nim [hunk 1]
```diff
@@ -1557,8 +1572,13 @@ template rawToFormalFileInfo(rawInfo, formalInfo): untyped =
     checkAndIncludeMode(S_IXOTH, fpOthersExec)
 
     formalInfo.kind = pcFile
-    if S_ISDIR(rawInfo.st_mode): formalInfo.kind = pcDir
-    if S_ISLNK(rawInfo.st_mode): formalInfo.kind.inc()
+    if S_ISDIR(rawInfo.st_mode):
+      formalInfo.kind = pcDir
+    elif S_ISLNK(rawInfo.st_mode):
+      # Symlinks can't occur for file handles so this if should never be
```
Comment (dom96) [inline lib/pure/os.nim]: This should be an ``assert`` then.

File: lib/pure/os.nim [hunk 2]
```diff
@@ -865,6 +865,19 @@ type
 
 {.deprecated: [TPathComponent: PathComponent].}
 
+
+when defined(posix):
+  proc getSymlinkFileKind(path: string): PathComponent =
+    # Helper function.
+    var s : Stat
```
Comment (dom96) [inline lib/pure/os.nim]: No spaces before colon please.

**Comment by dom96** [comment]
> I tried to pass nil instead of "" but then nim complained:
> ```Error: ambiguous call; both system.==(x: pointer, y: pointer) and system.==(x: seq[T], y: seq[T]) match for: (nil, nil)```

What line did you get this error on?

File: lib/pure/os.nim (hunk 2 above)
Comment (dom96) [inline lib/pure/os.nim]: Also, could you raise an error when ``path`` is ``""`` in this proc with a meaningful error?

File: lib/pure/os.nim (hunk 1 above)
Comment (rudis) [inline lib/pure/os.nim]: Will do.

File: lib/pure/os.nim (hunk 2 above)
Comment (rudis) [inline lib/pure/os.nim]: Is the additional check really necessary? It's only called with a normal path and the assert prevent other calls. I could add an assert if you like.

Comment (dom96) [inline lib/pure/os.nim]: Oh right. Nah, one assert is enough :)

**Comment by rudis** [comment]
It happens when I apply the following patch (ontop of my changes):

```diff
diff --git a/lib/pure/os.nim b/lib/pure/os.nim
index cde0ca4..cfa77ca 100644
--- a/lib/pure/os.nim
+++ b/lib/pure/os.nim
@@ -1575,7 +1575,7 @@ template rawToFormalFileInfo(rawInfo, path, formalInfo): untyped =
     if S_ISDIR(rawInfo.st_mode):
       formalInfo.kind = pcDir
     elif S_ISLNK(rawInfo.st_mode):
-      assert(path != "") # symlinks can't occur for file handles
+      assert(path != nil) # symlinks can't occur for file handles
       formalInfo.kind = getSymlinkFileKind(path)
 
 proc getFileInfo*(handle: FileHandle): FileInfo =
@@ -1592,12 +1592,12 @@ proc getFileInfo*(handle: FileHandle): FileInfo =
     var realHandle = get_osfhandle(handle)
     if getFileInformationByHandle(realHandle, addr rawInfo) == 0:
       raiseOSError(osLastError())
-    rawToFormalFileInfo(rawInfo, "", result)
+    rawToFormalFileInfo(rawInfo, nil, result)
   else:
     var rawInfo: Stat
     if fstat(handle, rawInfo) < 0'i32:
       raiseOSError(osLastError())
-    rawToFormalFileInfo(rawInfo, "", result)
+    rawToFormalFileInfo(rawInfo, nil, result)
 
 proc getFileInfo*(file: File): FileInfo =
   if file.isNil:
```

```
lib/pure/os.nim(1600, 24) template/generic instantiation from here
lib/pure/os.nim(1578, 19) template/generic instantiation from here
lib/system.nim(347, 10) Error: ambiguous call; both system.==(x: pointer, y: pointer) and system.==(x: cstring, y: cstring) match for: (nil, nil)
FAILURE

File: lib/pure/os.nim (hunk 2 above)
Comment (rudis) [inline lib/pure/os.nim]: Done.

**Comment by rudis** [comment]
Updated version pushed.

Regarding the tests: Do we have infrastructure for "proper" unit tests? Most tests I found just generate some output and check it. Is it okay to write tests like that or did I miss something?

**Comment by dom96** [comment]
Yes, that is how the majority of tests work. Others just check the exit code of the process. It's best to output something and verify it.

**Comment by dom96** [comment]
> It happens when I apply the following patch (ontop of my changes): [....]

Hrm, strange error.

**Comment by rudis** [comment]
Ok, will add some tests later.

I think it happens because `rawToFormalFileInfo` is template. I tried passing a `string` as second argument, but that didn't help. I don't know enough Nim yet to understand whats going on.

**Comment by dom96** [comment]
Okay, i'll merge it in for now then.

**Pull request merged**